### ISSUE TITLE
Deep linking and embed of annotations

### ISSUE THREAD

**Issue by anderspeders** [NRGI/resourcecontracts.org#375]
![image](https://cloud.githubusercontent.com/assets/1101585/10064092/fcfa9c1c-6242-11e5-8cc4-9ba734ff5d4c.png)

I would ssume that annotorious and annotator.js would enable deep linking to specific annotations quite easily - so that I can link to it in a document or embed it in an article?
Could you please estimate how heavy a lift this would be?

**Comment by anjesh** [comment]
I don't get this. Embed the annotation in a document/article? Is it something like embedding in the website using iframe kind of thing? Can you give some examples and usecase.

**Comment by jedm** [comment]
@anjesh This would be two things, as I see it. 

One is to make sure each annotation has a unique permanent URL that will open the relevant annotation and contract, similar to how OpenCorporates works for entities or Document Cloud works in our previous site version. (examples to follow below).

Another would be to create an EMBED function (not necessarily via iframe) so that a user could write a blog post about a contract or an annotation of a contract and include the specific data (contract and annotation) within their post. An embedded video is the obvious example of this but you also see it frequently with embedded data excerpts or with how Document Cloud enables contracts to be embedded in news articles online or blogs.

Finally, this same unique URL approach should be available for search results and (eventually) contract comparison results too.

I will post examples shortly in case it still is not clear.

**Comment by jedm** [comment]
Unique URLs
Open Corporates - KRAFT FOODS GROUP INC
https://opencorporates.com/companies/us_ma/363083135

RC.org/Document Cloud - COBAD - Guinée, 2001 contract, Local Employment Requirements, Art. 16, p. 14
https://www.documentcloud.org/documents/519538-convention-cobad.html#document/p14/a81656

Embed functions
Document Cloud/Wall Street Journal - Hillary Clinton's emails
http://graphics.wsj.com/hillary-clinton-emails/

Socrata
https://support.socrata.com/hc/en-us/articles/202950268-How-do-I-Embed-a-Dataset-or-Visualization-
The White House - records of WH visitors from Socrata database
https://www.whitehouse.gov/briefing-room/disclosures/visitor-records

Example of Unique URLs for comparisons:
EITI
https://eiti.org/countries/reports/compare/5838%2C4364%2C5877%2C3806%2C5879%2C4334

NRGI's Resource Governance Index - Compare Indicators for only Africa:
http://index.resourcegovernance.org/rgi/data-explore#!/?countries=1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17&indexes=0,2,14,36,53&externals=-1

**Comment by byndcivilization** [comment]
Let me jump in because im not sure if we're all on the same page. Jumping back up to Anders' question I think what he's referencing is the ability to link to a specific div/annotation by id. I think theres another name for that but cant remember off the top of my head. 

i.e. https://en.wikipedia.org/wiki/Taylor_Swift#Product_endorsements

However since we're operating in PDF you're not really dong this in divs or DOM elements really. I could be wrong but arent PDF viewers outside of the DOM? Even document cloud (which gets the closest) it isn't really linking to a specific DOM element, but rather calling a specific page template to render in a specific way.  From what I can see all the pdf linkages are fired by client side interactions but that all the data is passed on load. Is that right? Don't know if theres a non-rube goldberg way to fire off JS events via URI encoding. Maybe a jquery event fired by detecting url?

As far as unique contract IDs, I think thats tied to the OCID implementation that will happen but maybe after launch.  It should probs be something along the lines of "http://alpha.resourcecontracts.org/contract/<OCID>".

As far as embed. Yes embedding should come but not sure if its more important than getting a stable platform. I think iframe is a pretty typical way to go (both of the links jed sent use that and i think DC uses it also), i.e. you click a "generate embed" button and a dialog pops up that gives you the iframe code. 

With comparisons, I thought we jettisoned that until after launch. Either way I think comparison urls were already implemented...."http://alpha.openlandcontracts.org/contract/<OCID - 1>/<OCID - 2>/compare". That said I think overall we could do some work on our URL patterns to better follow our API conventions.

**Comment by jedm** [comment]
I have no opinion about the timing of an Embed featured, but all Annotations should have unique URLs. 

And all Annotations should have unique permanent URLs regardless of PDF or raw text. i.e., use the URL from anywhere and it takes you to the correct annotation in the correct contract.

I thought that each contract already had a unique permanent URL. Is this not the case?

**Comment by byndcivilization** [comment]
I dont think this is possible with PDFs in the way that we're doing it, but we should wait for YI to chime in. This is the problem with trying to do this with both PDFs and Text. 

Sorry the links seem to be removing then ends of the above urls. We have unique ids in that they are the DB IDs. We've been waiting on getting OCID sorted before issueing specific permanent IDs. Weve now got auth to issue OCIDs, guidance on format, and a schema to implement so that should be done soon.

**Comment by jedm** [comment]
@byndcivilization that makes sense re PDFs. I assume that if that problem were solved then the continuing problem of linking the left column (annotations) and the center column (page viewer w/PDF or text) could also be solved.

However, I assume that a URL with " #[string] " appended CAN go to the correct contract and scroll the left column to the correct annotation. Please confirm.

**Comment by byndcivilization** [comment]
theres this.
http://alpha.resourcecontracts.org/contract/876#state-participationparticipation-de-letat
its not explicitly given but it should be doable for @anjesh  to add an anchor link.

On the view page YI dont use Id's so not sure how much work it would take. Also not sure how anchor tagging works on multipanel viewports.

**Comment by jedm** [comment]
@byndcivilization that works really well. If, for now, the "direct" links would go to summary page not Contract page, should we add a small link in each summary page annotation (e.g., the link you just posted) back to Contract view? 

And yeah, seems even more important to know if there's some kind of syntax/control for multipanel views that was like [leftcolpos7&centerview1pos7] or something like that.

**Comment by byndcivilization** [comment]
there might be (and ill let @anjesh chime in) but you're opening a can of worms in terms of x-browser compatiblity.

**Comment by anjesh** [comment]
We now have link for the annotations, which can be shared. If we are looking for embed feature, lets discuss in new issue.
